peekingduck/pipeline/nodes/model/jde_mot/jde_files/multitracker.py

```diff
@@ -0,0 +1,546 @@
+        self.removed_stracks = []  # type: List[STrack]
+
+        self.frame_id = 0
+        self.det_thresh = opt["conf_threshold"]
+        self.buffer_size = int(frame_rate / 30.0 * opt["track_buffer"])
+        self.max_time_lost = self.buffer_size
+
+        self.kalman_filter = KalmanFilter()
+
+    def update(self, im_blob: torch.float32, img0: np.ndarray) -> List[STrack]:
+        """Processes the image frame and finds bounding box(detections).
+        Associates the detection with corresponding tracklets and also
+        handles lost, removed, refound and active tracklets
+
+        Args:
+            im_blob (torch.float32): Tensor of shape depending upon the
+                size of image. By default, shape of this tensor is [1, 3, 608, 1088]
+            img0 (np.ndarray): ndarray of shape depending on the input
+                image sequence. By default, shape is [608, 1080, 3]
+
+        Returns:
+            List[STrack]: The list contains instance information regarding
+                the online_tracklets for the recieved image tensor.
+        """
+        self.frame_id += 1
+        activated_starcks = []  # for storing active tracks, for the current frame
+        refind_stracks = []
+        # Lost Tracks whose detections are obtained in the current frame
+        lost_stracks = []
+        # The tracks which are not obtained in the current frame but are
+        # not removed. (Lost for some time lesser than the threshold for removing)
+        removed_stracks = []
+
+        """ Step 1: Network forward, get detections & embeddings"""
+        with torch.no_grad():
+            pred = self.model(im_blob)
+        # pred is tensor of all the proposals (default number of proposals: 54264).
+        # Proposals have information associated with the bounding box and embeddings.
+        pred = pred[pred[:, :, 4] > self.opt["conf_threshold"]]
+        # pred now has lesser number of proposals.
+        # Proposals rejected on basis of object confidence score.
+        if len(pred) > 0:
+            dets = non_max_suppression(
+                pred.unsqueeze(0), self.opt["conf_threshold"], self.opt["nms_threshold"]
+            )[0].cpu()
+            # Final proposals are obtained in dets. Information of bounding box
+            # and embeddings also included. Next step changes the detection scales
+            scale_coords(self.opt["img_size"], dets[:, :4], img0.shape).round()
+            """Detections is list of (x1, y1, x2, y2, object_conf, class_score, class_pred)"""
+            # class_pred is the embeddings.
+            detections = [
+                STrack(STrack.tlbr_to_tlwh(tlbrs[:4]), tlbrs[4], f.numpy(), 30)
+                for (tlbrs, f) in zip(dets[:, :5], dets[:, 6:])
+            ]
+        else:
+            detections = []
+
+        """ Add newly detected tracklets to tracked_stracks"""
+        unconfirmed = []
+        tracked_stracks = []  # type: List[STrack]
```

Comment:
prefer to type directly instead of using comment

Reply:
Updated